### PR TITLE
Migrate Image Training Endpoint to Django

### LINKED ISSUE #993: [FEATURE]: Migrate Image Training Endpoint
### Feature Name

Migrate Image Training Endpoint`

### Your Name

Karthik Subramanian

### Description

DLP is migrating the training endpoints from Flask Blueprint to Django. This task requires creation of `/image` route to run image training requests in Django. 

References/Resources:
* For structure, mirror it similar to [here](https://github.com/DSGT-DLP/Deep-Learning-Playground/blob/362ee4594c18a2363922b392379fc73388712a9f/training/training/routes/tabular/tabular.py#L17-L65). 
* Note that image train doesn't support regression. Focus on classification case here!
* Old image train code for inspiration: [here](https://github.com/DSGT-DLP/Deep-Learning-Playground/blob/nextjs/backend/dl/dl_trainer.py#L288-L414)

Please ensure that this endpoint is tested on a sample model/architecture for torch built in dataset (eg: MNIST, FashionMNIST, etc) to ensure that the model is training properly

### PR DESCRIPTION
**Pull request by codingwithsurya**
# Migrate Image Training Endpoint to Django

Github Issue Number Here: Feature#993

**What user problem are we solving?**

With the ongoing migration from Flask Blueprint to Django for better scalability and maintainability, we need to ensure that the training endpoints for image datasets continue to function as expected. Users need a reliable endpoint to submit image training requests and this migration aims to improve the overall infrastructure.

**What solution does this PR provide?**

This PR introduces the /image route within the Django framework, replacing the old Flask Blueprint route. The new route handles image training requests, particularly focusing on classification tasks as regression is not supported. The structure of this endpoint is mirrored from the specified reference, with adjustments made to fit the Django framework. The old image train code served as inspiration for ensuring that the necessary features and behaviors are retained.

**Testing Methodology**

Testing yet to be done but to verify the functionality and integrity of the new /image route, but when testing begins, we can submit training requests using torch built-in datasets (MNIST, FashionMNIST) to ensure that models train as expected and the endpoint returns accurate results.

Also use postman for testing .

**Any other considerations**

### CODE REVIEW

**Comment by codingwithsurya** [comment]
@karkir0003 @dwu359 

took a stab at this. a couple questions i had was. 

1. What should the correct path be in ImageDefaultDatasetCreator constructor (dataset.py)? Here is the code I am referring to:

`self.train_set = datasets.__dict__[dataset_name](root='./backend/image_data_uploads', train=True, download=True, transform=self.train_transform)`

`self.test_set = datasets.__dict__[dataset_name](root='./backend/image_data_uploads', train=False, download=True, transform=self.test_transform)`

With transition to django, is the path for data uploads still gonna be in backend/data_uploads?

2. Does my image.py look good in /routes/image? Is there any particular i have to add for image training?

3. Any additional schemas for image classification? I think i got the right ones. 


If everything looks good here (no red flags in code), i can begin testing with our default datsets to make sure our trainspace flow is working as intended.

**Comment by karkir0003** [comment]
for image data uploads, please dont use backend/ dir.

File: training/training/core/dataset.py [hunk 1]
```diff
@@ -98,3 +101,75 @@ def getCategoryList(self) -> list[str]:
         if self._category_list is None:
             raise Exception("Category list not available")
         return self._category_list
+
+
+class DefaultImageDatasets(Enum):
+    MNIST = "MNIST"
+    FASHION_MNIST = "FashionMNIST"
+    KMNIST = "KMNIST"
```
Comment (karkir0003) [inline training/training/core/dataset.py:111]: is this real? @codingwithsurya

File: training/training/core/dataset.py [hunk 2]
```diff
@@ -98,3 +101,75 @@ def getCategoryList(self) -> list[str]:
...
+
+
+class ImageDefaultDatasetCreator(TrainTestDatasetCreator):
+    def __init__(
+        self,
+        dataset_name: str,
+        train_tarnsform: None,
```
Comment (karkir0003) [inline training/training/core/dataset.py]: misspell

File: training/training/core/dataset.py [hunk 3]
```diff
@@ -98,3 +101,75 @@ def getCategoryList(self) -> list[str]:
...
+        test_transform: None,
+        batch_size: int = 32,
+        shuffle: bool = True,
+    ):
+        if dataset_name not in DefaultImageDatasets.__members__:
+            raise Exception(
+                f"The {dataset_name} file does not currently exist in our inventory. Please submit a request to the contributors of the repository"
+            )
+        self.train_transform = train_tarnsform or transforms.Compose(
```
Comment (karkir0003) [inline training/training/core/dataset.py]: misspell

File: training/training/core/dataset.py [hunk 4]
```diff
@@ -98,3 +101,75 @@ def getCategoryList(self) -> list[str]:
...
+            [transforms.toTensor()]
+        )
+        self.test_transform = test_transform or transforms.Compose(
+            [transforms.toTensor()]
+        )
+        self.batch_size = batch_size
+        self.shuffle = shuffle
+
+        self.train_set = datasets.__dict__[dataset_name](
```
Comment (karkir0003) [inline training/training/core/dataset.py:146]: idk if this works. can u verify in ipython

**Comment by codingwithsurya** [comment]
> for image data uploads, please dont use backend/ dir.

sure, then what directory do i use? I dont see a similar directory in /training

File: training/training/core/dataset.py [hunk 5]
```diff
@@ -98,3 +101,75 @@ def getCategoryList(self) -> list[str]:
...
+
+    def createTrainDataset(self) -> DataLoader:
+        return DataLoader(
+            self.train_set,
+            batch_size=self.batch_size,
+            shuffle=self.shuffle,
+            drop_last=True,
+        )
+
+    def createTestDataset(self):
```
Comment (karkir0003) [inline training/training/core/dataset.py]: put return typing using -> syntax. this is standard for python3

File: training/training/routes/image/image.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        )
+
+        test_loader = DataLoader(
+            dataCreator.createTestDataset(),
+            batch_size=imageParams.batch_size,
+            shuffle=False,
+            drop_last=True,
+        )
+
+        model = DLModel.fromLayerParamsList(imageParams.user_arch)
```
[the remark this answers is not in the document]
Comment (karkir0003) [reply training/training/routes/image/image.py:25]: @codingwithsurya straightforward fix?

[the remark this answers is not in the document]
Comment (karkir0003) [reply training/training/routes/image/image.py:25]: should be straightforward to fix @codingwithsurya

File: training/training/routes/image/image.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+@router.post("", auth=FirebaseAuth())
+def imageTrain(request: HttpRequest, imageParams: ImageParams):
+    if imageParams.default:
+        dataCreator = ImageDefaultDatasetCreator.fromDefault(
+            imageParams.default, imageParams.test_size, imageParams.shuffle
+        )
+        train_loader = DataLoader(
+            dataCreator.createTrainDataset(),
+            batch_size=imageParams.batch_size,
+            shuffle=False,
```
Comment (karkir0003) [inline training/training/routes/image/image.py]: is shuffle an arg the user can specify @farisdurrani ?

File: training/training/core/dataset.py [hunk 8]
```diff
@@ -12,6 +12,9 @@
 import torch
 from torch.utils.data import Dataset
 from torch.autograd import Variable
+from torchvision import datasets, transforms
+from torch.util.data import DataLoader, random_split
```
[the remark this answers is not in the document]
Comment (karkir0003) [reply training/training/core/dataset.py:16]: check this import @codingwithsurya

**Review by karkir0003 — COMMENTED** [review]
left comments

File: training/training/routes/image/image.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (codingwithsurya) [reply training/training/routes/image/image.py:25]: i think i can just fix this by importing image. if not, i'l try using the Sequence thing they suggested.

[the remark this answers is not in the document]
Comment (codingwithsurya) [reply training/training/routes/image/image.py:25]: i think i can just fix this by importing image. if not, i'l try using the Sequence thing they suggested.

File: training/training/core/dataset.py (hunk 1 above)
Comment (codingwithsurya) [reply training/training/core/dataset.py:111]: yea i imported torchvision datasets

File: training/training/core/dataset.py (hunk 5 above)
Comment (codingwithsurya) [reply training/training/core/dataset.py]: oops, fixed now

File: training/training/core/dataset.py (hunk 3 above)
Comment (codingwithsurya) [reply training/training/core/dataset.py]: oops, fixed now

File: training/training/core/dataset.py (hunk 2 above)
Comment (codingwithsurya) [reply training/training/core/dataset.py]: oops, fixed now

File: training/training/core/dataset.py (hunk 4 above)
Comment (codingwithsurya) [reply training/training/core/dataset.py:146]: Will do.

File: training/training/core/dataset.py (hunk 1 above)
Comment (karkir0003) [reply training/training/core/dataset.py:111]: i see. so KMNIST is an actual dataset. interesting. What about the CIFAR dataset? can u add that

File: training/training/routes/image/image.py (hunk 7 above)
Comment (karkir0003) [reply training/training/routes/image/image.py]: @codingwithsurya looks like shuffle is an arg the user can specify.

File: training/training/core/dataset.py [hunk 9]
```diff
@@ -98,3 +101,75 @@ def getCategoryList(self) -> list[str]:
...
+            [transforms.toTensor()]
+        )
+        self.test_transform = test_transform or transforms.Compose(
+            [transforms.toTensor()]
+        )
+        self.batch_size = batch_size
+        self.shuffle = shuffle
+
+        self.train_set = datasets.__dict__[dataset_name](
+            root="./backend/image_data_uploads",
```
Comment (karkir0003) [inline training/training/core/dataset.py]: please use ./training/image_data_uploads but ensure that after dataloader created the downloaded dataset deleted. might wanna confirm with torch docs here @codingwithsurya

**Review by karkir0003 — COMMENTED** [review]
added more comments @codingwithsurya

**Comment by karkir0003** [comment]
i think u can test the endpoint out @codingwithsurya

File: training/training/core/dataset.py (hunk 1 above)
Comment (codingwithsurya) [reply training/training/core/dataset.py:111]: sure

File: training/training/routes/image/image.py (hunk 7 above)
Comment (codingwithsurya) [reply training/training/routes/image/image.py]: sounds good.

File: training/training/core/dataset.py (hunk 9 above)
Comment (codingwithsurya) [reply training/training/core/dataset.py]: Ok will create that directory. And u would want a similar structure as in backend/image_data_uploads? right? but obviously in that django blueprint.

Comment (karkir0003) [reply training/training/core/dataset.py]: yes

Comment (karkir0003) [reply training/training/core/dataset.py]: I've DM'd you on discord this detail

Comment (codingwithsurya) [reply training/training/core/dataset.py]: sure, check out this gist on what im thinking we can do:

https://gist.github.com/codingwithsurya/950097788ac8f60b561d62ca087fed5f

Comment (karkir0003) [reply training/training/core/dataset.py]: we discussed over dm on this @codingwithsurya

**Comment by karkir0003** [comment]
> > for image data uploads, please dont use backend/ dir.
> 
> sure, then what directory do i use? I dont see a similar directory in /training

should be clarified

**Comment by codingwithsurya** [comment]
> > > for image data uploads, please dont use backend/ dir.
> > 
> > 
> > sure, then what directory do i use? I dont see a similar directory in /training
> 
> should be clarified

yea looks good. i just commited what we talked about. and i'll test on colab as well soon when i get a chance.

File: training/training/routes/image/image.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+                test_loader,
+                model,
+                optimizer,
+                criterionHandler,
+                imageParams.epochs,
+                dataCreator.getCategoryList(),
+            )
+            for epoch_result in trainer:
+                print(epoch_result)
+            print(trainer.labels_last_epoch, trainer.y_pred_last_epoch)
```
Comment (farisdurrani) [inline training/training/routes/image/image.py:41]: Are all these print statements necessary?

Comment (codingwithsurya) [reply training/training/routes/image/image.py:41]: saw this in the original tabular.py as well. if they're not necessary, i can remove them.

File: training/training/core/dataset.py [hunk 11]
```diff
@@ -98,3 +103,90 @@ def getCategoryList(self) -> list[str]:
...
+            train=False,
+            download=True,
+            transform=self.test_transform,
+        )
+
+    @classmethod
+    def fromDefault(
+        cls,
+        dataset_name: str,
+        train_transform=None,
```
Comment (karkir0003) [inline training/training/core/dataset.py:163]: add typing here

File: training/training/core/dataset.py [hunk 12]
```diff
@@ -98,3 +103,90 @@ def getCategoryList(self) -> list[str]:
...
+        test_transform=None,
```
Comment (karkir0003) [inline training/training/core/dataset.py:164]: same here

File: training/training/routes/image/__init__.py [hunk 13]
```diff
@@ -0,0 +1,5 @@
+from training.routes.image.image import router
+
+
+def get_image_router():
```
Comment (karkir0003) [inline training/training/routes/image/__init__.py]: add return typing

Comment (karkir0003) [reply training/training/routes/image/__init__.py]: using the `-> <return type>` syntax

**Review by karkir0003 — COMMENTED** [review]
added minor comments on static typing python

File: training/training/core/dataset.py (hunk 4 above)
Comment (codingwithsurya) [reply training/training/core/dataset.py:146]: Screenshot 2023-10-12 at 8 39 21 PM

works on ipython 👍🏽 @karkir0003

File: training/training/routes/image/image.py (hunk 10 above)
Comment (codingwithsurya) [reply training/training/routes/image/image.py:41]: @farisdurrani @karkir0003 any clarification on what to do here. Are all the print statements necessary in grand scheme of things?

File: training/training/core/dataset.py (hunk 11 above)
Comment (codingwithsurya) [reply training/training/core/dataset.py:163]: done!

**Comment by codingwithsurya** [comment]
> added minor comments on static typing python

addressed all these

**Comment by codingwithsurya** [comment]
TO-DO: find viable nnSequential PyTorch MNIST Architecture for testing

File: training/training/core/dataset.py (hunk 9 above)
Comment (codingwithsurya) [reply training/training/core/dataset.py]: implemented. Testing will confirm if this truly works.

**Comment by codingwithsurya** [comment]
TODO: Cross-Entropy Loss not computing correctly. Figure out correct param for FLATTEN in current architecture. 

Otherwise, we can make our own

**Comment by codingwithsurya** [comment]
Update: MNIST dataset testing works!

TODO: Train/test transforms as parameter to endpoint. CIFAR10 Testing.

File: training/training/routes/image/image.py (hunk 10 above)
Comment (karkir0003) [reply training/training/routes/image/image.py:41]: did we resolve this @codingwithsurya
